### PR TITLE
COR-514-excel-upload

### PR DESCRIPTION
**Pull request by neb42**
* Support excel and csv uploads in handler
* Add individual excel unmarshal
* Map file column names to db column names (currently using the same names, but allows us to differ)

### CODE REVIEW

**Comment by ludydoo** [comment]
Do you want to add tests in another PR ?

**Review by internetti — COMMENTED** [review]
when trying to upload the bla.xlsx I get this error: `failed to put records: sql: Scan error on column index 15, name "country_id": converting NULL to string is unsupported`
will this be fixed later?

File: internal/api/individual_tabular.go [hunk 1]
```diff
@@ -2,43 +2,78 @@ package api
...
+	f, err := excelize.OpenReader(reader)
+
+	if err != nil {
+		return err
+	}
+
+	// defer func() {
+	// 	if err := f.Close(); err != nil {
+	// 	}
+	// }()
```
Comment (ludydoo) [inline internal/api/individual_tabular.go:36]: Unused ?

File: internal/api/individual_tabular.go [hunk 2]
```diff
@@ -2,43 +2,78 @@ package api
...
 	csvReader.TrimLeadingSpace = true
 	records, err := csvReader.ReadAll()
 	if err != nil {
-		return nil, err
+		return err
+	}
+	return unmarshalIndividualsTabularData(records, individuals, fields)
+}
+
+func UnmarshalIndividualsExcel(reader io.Reader, individuals *[]*Individual, fields *[]string) error {
```
Comment (ludydoo) [inline internal/api/individual_tabular.go:26]: I don't think you need the `fields` parameter here.

**Comment by neb42** [comment]
> when trying to upload the bla.xlsx I get this error: `failed to put records: sql: Scan error on column index 15, name "country_id": converting NULL to string is unsupported` will this be fixed later?

This is the country id issue I mentioned in standup today. We used to upload it as part of the excel file, but now it is done via the query string. It was broke in a different pr and I didn't think it was part of this pr to fix it.

File: internal/api/individual_tabular.go (hunk 2 above)
Comment (neb42) [reply internal/api/individual_tabular.go:26]: I think I do. This is using the same partial upload logic that you added for csv uploads

File: internal/constants/individual.go [hunk 3]
```diff
@@ -44,3 +60,75 @@ const (
 	FileColumnIndividualPresentsProtectionConcerns = "presents_protection_concerns"
 	FileColumnIndividualSensoryImpairment          = "sensory_impairment"
 )
+
+var IndividualDBColumns = []string{
```
Comment (ludydoo) [inline internal/constants/individual.go:64]: Unused variable ?

File: internal/constants/individual.go [hunk 4]
```diff
@@ -28,6 +28,22 @@ const (
...
+	DBColumnIndividualFullName                   = "full_name"
+	DBColumnIndividualGender                     = "gender"
+	DBColumnIndividualIsMinor                    = "is_minor"
+	DBColumnIndividualMentalImpairment           = "mental_impairment"
+	DBColumnIndividualPhoneNumber                = "phone_number"
+	DBColumnIndividualPhysicalImpairment         = "physical_impairment"
+	DBColumnIndividualPreferredName              = "preferred_name"
+	DBColumnIndividualPresentsProtectionConcerns = "presents_protection_concerns"
+	DBColumnIndividualSensoryImpairment          = "sensory_impairment"
+
```
Comment (ludydoo) [inline internal/constants/individual.go:46]: Not sure about the `DBColumn...` name, since it's not used in the DB at all. But we can check it later

**Comment by neb42** [comment]
I'm not going to add tests for `UnmarshalIndividualsExcel` and `UnmarshalIndividualsCSV` as they are fairly simple, harder to test, and could be better handled by some integration tests once we're ready.

**Pull request merged**